images/handlers.go

```diff
@@ -87,21 +87,19 @@ func Handlers(handlers ...Handler) HandlerFunc {
 // This differs from dispatch in that each sibling resource is considered
 // synchronously.
 func Walk(ctx context.Context, handler Handler, descs ...ocispec.Descriptor) error {
-	for _, desc := range descs {
-
+	queue := descs
+	for len(queue) > 0 {
+		desc := queue[0]
+		queue = queue[1:]
 		children, err := handler.Handle(ctx, desc)
 		if err != nil {
 			if errors.Is(err, ErrSkipDesc) {
-				continue // don't traverse the children.
+				// don't traverse children.
+				continue
 			}
 			return err
 		}
-
-		if len(children) > 0 {
-			if err := Walk(ctx, handler, children...); err != nil {
-				return err
-			}
-		}
+		queue = append(queue, children...)
```

Comment:
Sorry, I did not use this piece of specific testing, according to my personal point of view this will be faster, this piece I can not directly write benchmark, if you think the append consumption will be larger, you can use `container/list`, but the code will be a little more complex, what do you think @fuweid

Reply:
```go
func Walk(ctx context.Context, handler Handler, descs ...ocispec.Descriptor) error {
	queue := list.New()
	for _, desc := range descs {
		queue.PushBack(desc)
	}

	for queue.Len() > 0 {
		elem := queue.Front()
		desc := elem.Value.(ocispec.Descriptor)
		queue.Remove(elem)

		children, err := handler.Handle(ctx, desc)
		if err != nil {
			if errors.Is(err, ErrSkipDesc) {
				// don't traverse children.
				continue
			}
			return err
		}
		for _, child := range children {
			queue.PushBack(child)
		}
	}
	return nil
}
```

Reply:
Hi @helen-frank 

To be honest, I'm not expert on stack cost. Based on the real case, the image should have 3 level at most.

> This specification defines an OCI Image, consisting of an [image manifest](https://github.com/opencontainers/image-spec/blob/main/manifest.md), an [image index](https://github.com/opencontainers/image-spec/blob/main/image-index.md) (optional), a set of [filesystem layers](https://github.com/opencontainers/image-spec/blob/main/layer.md), and a [configuration](https://github.com/opencontainers/image-spec/blob/main/config.md).
>
> From https://github.com/opencontainers/image-spec/blob/main/spec.md

```
0000000000ef4760 <github.com/containerd/containerd/images.Walk@@Base>:
  ef4760:       4c 8d a4 24 48 ff ff    lea    -0xb8(%rsp),%r12
  ef4767:       ff
  ef4768:       4d 3b 66 10             cmp    0x10(%r14),%r12
  ef476c:       0f 86 16 02 00 00       jbe    ef4988 <github.com/containerd/containerd/images.Walk@@Base+0x228>
  ef4772:       48 81 ec 38 01 00 00    sub    $0x138,%rsp
  ef4779:       48 89 ac 24 30 01 00    mov    %rbp,0x130(%rsp)
```

`0x138` stack size and one descriptor size is around 120 bytes.
When `append` needs to re-allocate the space, it seems like that stack might be good than this change.

In my opinion, there are 3 level at most, which is acceptable and the code is clean and easy to understand.
I want to keep it as it is.

Reply:
Got it, learned something new